src/pubsub.js

```diff
@@ -0,0 +1,586 @@
+                                  .length === 0
+            if (hasAllPeers) {
+              clearInterval(i)
+              callback()
+            }
+          })
+        }, 1000)
+      }
+
+      describe('.publish', () => {
+        it('message from string', (done) => {
+          ipfs1.pubsub.publish(topic, 'hello friend', done)
+        })
+
+        it('message from buffer', (done) => {
+          ipfs1.pubsub.publish(topic, new Buffer('hello friend'), done)
+        })
+      })
+
+      describe('.subscribe', () => {
+        it('to one topic', (done) => {
+          ipfs1.pubsub.subscribe(topic, (err, subscription) => {
+            expect(err).to.not.exist
+
+            subscription.on('data', (msg) => {
+              expect(msg.data).to.equal('hi')
+              subscription.cancel(done)
+            })
+
+            ipfs1.pubsub.publish(topic, 'hi', (err) => {
+              expect(err).to.not.exist
+            })
+          })
+        })
+
+        it('errors on double subscription', (done) => {
+          series([
+            (cb) => ipfs1.pubsub.subscribe(topic, cb),
+            (cb) => ipfs1.pubsub.subscribe(topic, cb)
+          ], (err, subs) => {
+            expect(err).to.exist
+            expect(err.toString())
+              .to.eql(`Error: Already subscribed to '${topic}'`)
+            subs[0].cancel(done)
+          })
+        })
+
+        it('discover options', (done) => {
+          ipfs1.pubsub.subscribe(topic, {
+            discover: true
+          }, (err, subscription) => {
+            expect(err).to.not.exist
+            subscription.cancel(done)
+          })
+        })
+      })
+
+      describe('subscription', () => {
+        it('.cancel and wait for callback', (done) => {
+          ipfs1.pubsub.subscribe(topic, (err, subscription) => {
+            expect(err).to.not.exist
+            subscription.cancel(done)
+          })
+        })
+
+        it('.cancel and wait for end event', (done) => {
+          ipfs1.pubsub.subscribe(topic, (err, subscription) => {
+            expect(err).to.not.exist
+            subscription.on('end', done)
+            subscription.cancel()
+          })
+        })
+      })
+
+      describe('.peers', () => {
+        before((done) => {
+          ipfs2.id((err, id) => {
+            expect(err).to.not.exist
+            const ipfs2Addr = id.addresses[0]
+            ipfs1.swarm.connect(ipfs2Addr, (err) => {
+              expect(err).to.not.exist
+              // We need to fix this on libp2p level
+              setTimeout(done, 3000)
+            })
+          })
+        })
+
+        // TODO clarify what is the goal of pubsub.peers
+        it('returns an error when not subscribed to a topic', (done) => {
+          ipfs1.pubsub.peers(topic, (err, peers) => {
+            expect(err).to.exist
+            expect(err.toString()).to.equal(`Error: Not subscribed to '${topic}'`)
+            done()
+          })
+        })
```

Comment:
Well, for me `peers(topic)` tells me I should expect "give me the peers I have in my peerSet that are subscribed to a given topic".